core/lib/src/config/custom_values.rs

```diff
@@ -42,6 +43,47 @@ impl fmt::Display for SecretKey {
     }
 }
 
+#[derive(Copy, Clone)]
+pub enum Port {
+    Generated(u16),
+    Provided(u16)
+}
+
+impl Deref for Port {
+    type Target = u16;
+
+    fn deref(&self) -> &u16 {
+        match self { Port::Generated(v) | Port::Provided(v) => v }
+    }
+}
+
+impl DerefMut for Port {
+    fn deref_mut(&mut self) -> &mut u16 {
+        match self { Port::Generated(v) | Port::Provided(v) => v }
+    }
+}
+
+impl Port {
+    crate fn value(&self) -> u16 {
+        match self {
+            Port::Generated(v) | Port::Provided(v) => *v
```

Comment:
We shouldn't need the `*v`. Instead, take `self` by value (`self` instead of `&self`). This is just as cheap as what's happening now.